crates/bevy_transform/src/systems.rs

```diff
@@ -47,50 +47,84 @@ pub fn sync_simple_transforms(
-                        &parent_query,
-                        child,
-                        changed || actual_parent.is_changed(),
-                    );
+    // Optimistically work on each root in parallel. The following algorithm propagates transforms down
+    // all hierarchies with changed or orphaned roots. if the root node itself is unchanged, this will not
+    // recurse into it's descendants.
+    root_query.par_iter().for_each(|(root, transform)| {
+        // Abort if this root node has not been meaningfully changed.
+        if !transform.is_changed() && orphaned_entities.binary_search(&root).is_err() {
+            return;
+        }
+        // SAFETY:
+        // - `root`'s ancestors are vacuously consistent (it has none).
+        // - We may operate as if all descendants are consistent, since `propagate_recursive` will panic before
+        //   continuing to propagate if it encounters an entity with inconsistent parentage.
+        // - Since each root entity is unique and the hierarchy is consistent and forest-like,
+        //   other root entities' `propagate_recursive` calls will not conflict with this one.
+        // - `transform_query` has not yet been used, so there can be no conflicting fetches elsewhere.
+        unsafe {
+            propagate_recursive(
+                &GlobalTransform::IDENTITY,
+                &transform_query,
+                &parent_query,
+                root,
+            );
+        };
+    });
+    // Optimistically work on each changed entity in parallel. The following algorithm finds minimal
+    // disjoint sub-trees in hierarchies without changed or orphaned roots, and updates them in parallel.
+    changed.par_iter().for_each(|entity| {
+        // Abort if this is a root node. This case is handled separately.
+        let Ok((_, parent)) = parent_query.get(entity) else {
+            return;
+        };
+        // Abort if the ancestors of this entity also have changed transforms or are orphaned.
+        let mut current = entity;
+        loop {
+            let Ok((_, current_parent)) = parent_query.get(current) else {
+                if orphaned_entities.binary_search(&current).is_ok() {
+                    return;
                 }
+                break;
+            };
+            current = current_parent.get();
+            if changed.contains(current) {
+                return;
             }
-        },
-    );
+        }
+        // Determine the global transform of the parent.
+        let mut parent_transform = &GlobalTransform::IDENTITY;
+        if let Ok((_, global_transform, _)) = transform_query.get(parent.get()) {
+            parent_transform = global_transform;
+        }
+        // SAFETY:
+        // Define any changed entity that is not a descendant of another changed entity to be an 'entry point'.
+        // - Since the hierarchy has forest structure, two distinct entry points cannot have shared decedents.
```

Comment:
FYI this is not entirely true because you could have cycles, but luckily they get filtered out by the loop above that looks for a changed ancestor.

In general though I feel like some of these safety comments are a bit generous with the assumptions.

Reply:
Maybe I should have marked this as a draft. I know I was a bit too informal with these, but I'm still getting used to the rust safety-comment proof style so I figured I could use the feedback.